cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,473 @@
+
+This diagram shows the way the API flows work. It’s assumed before we start that SPIFFE identities for the source and destination have already been created in the SPIRE server.
+
+![SPIFFE and SPIRE Cilium Agent using Delegated Identity API](./images/SPIFFE-SPIRE-cilium-agent-delegated.png)
+
+Then, when required (whether we do this at Endpoint creation or later is still TBD):
+1. All cilium-agents running a workload mentioned in an auth-enabled policy connect to their local spire-agent, and watch a set of SPIFFE labels that includes the labels for the relevant SVID. The cilium-agent identity is allowed to do Delegated Identity requests to the Workload API (WL API), so this is allowed.
+1. The spire-agent watches SVIDs on the SPIRE server, sees any updates to all relevant SVIDs, and passes them back to the cilium-agent.
+1. The cilium-agent then figures out where the request is coming from, and connects to the cilium-agent on the source node to perform the mTLS handshake. Because this  a mutual TLS, if it succeeds, then the workloads are authenticated.
+The destination node will also record that the workload is authenticated for reverse traffic.
+1. The mTLS handlers in each cilium-agent then pass the auth success to their local dataplane, by telling it that identity A is authenticated to identity B for some period of time (the lifetime of the certificates in the mTLS handshake).
+The mTLS handlers will also need to indicate the remote node, along with the auth type.
+For that lifetime, traffic between Identity A and Identity B will flow without further checks.
+
+#### SPIFFE installation steps and flow
+
+1. User creates a cluster, without CNI, per usual Cilium install process. 
+1. User installs Cilium, with mTLS enabled, SPIFFE auth will not work until the following conditions are met:
+    * SPIRE Server must be installed somewhere and configured correctly. (Bring-your-own-SPIRE-Server is an anticipated future need, but the initial implementation will use an in-cluster SPIRE server.)
+    * Per-node SPIRE Agent must be installed in the cluster and configured correctly. The SPIRE agent talks to the SPIRE server over the network, but all other communication with the Cilium Agent is via domain sockets shared on the host's filesystem.
```

Comment:
This is all accurate to my understanding of our previous discussions, though I'll note that I'm assuming that:
- SPIRE Server would be an independent helm deployment
- SPIRE Agent would be integrated into the Cilium helm charts as an additional sidecar/container.

(Just being explicit here since the text wasn't entirely clear on this, though again the text is fine, it's prescribing what needs to be in place, then it's up to the implementation to provide these in some way)

Reply:
Though all that said, I guess it's also a reasonable option to independently deploy the SPIRE agents on the nodes, that part doesn't necessarily need to be embedded in the Cilium daemonset. I guess we should be having this discussion on the linked issue rather than in the CFP ;)

Reply:
Haha yeah, I'm hoping to dig into the installation more this week.